### ISSUE TITLE
What framework to use? 

### ISSUE THREAD

**Issue by pietervdvn** [pietervdvn/MapComplete#1142]
If we consider a framework to switch to, the following questions should be answered:

- Can the framework work side-by-side with the current homegrown framework? This way, the codebase can be migrated step-by-step.
  - Can it insert HTML-elements?
  - Can it do (virtual) routing requests? (Not critical though) 
  - Can it adapt to the 'UIEventSource'-flow?
- How are translations handled?
  - Including RTL-languages?
  - Is loading a translation dynamically (from a theme definition) possible?
- How good are the accessibility-features (e.g. screenreaders)? Can it be made [WCAG](https://www.w3.org/WAI/standards-guidelines/wcag/)-compatible?
- How good is the tooling around the framework? Does it have a special language or is it plain typescript?
- What support does it have for MapLibre? Is a package available?

**Comment by pietervdvn** [comment]
@tordans Feel free to add some more important questions and/or to pitch some frameworks.

**Comment by wjtje** [comment]
Maybe you can take a look at svelte, I already got a version of MapComplete working with svelte that can build in less that 20 seconds (including loading from the browser).

https://user-images.githubusercontent.com/33957974/204669143-e220ce32-680c-45b9-801e-5ab5ffc9e27a.mp4

> Can the framework work side-by-side with the current homegrown framework?

Yes, you can create svelte element outside of svelte.

```ts
import InputElement from "./InputElement.svelte"

new InputElement({
    target: document.getElementById("target"),
    props: {
        title: "Name",
    },
})
```
```svelte
<script lang="ts">
  export let title: string;
</script>

<h1>{title}</h1>
```

> Can it do (virtual) routing requests? (Not critical though)

Yes using sveltekit.

> How are translations handled? 

You can use a package like svelte-i18n

> How good are the accessibility-features (e.g. screenreaders)?

I would say quite good, you have almost direct control over the html that will end up in the dom. Svelte will even try to help some of the common mistakes, https://svelte.dev/docs#accessibility-warnings .

> How good is the tooling around the framework? Does it have a special language or is it plain typescript?

It uses default typescript, and there are extensions for the most common used editors. But the syntax of svelte is so close to normal html, that you can easily use it without any extensions.

> What support does it have for MapLibre? Is a package available?

There are multiple ways of using MapLibre with svelte.

**Comment by Flaburgan** [comment]
Svelte is the currently "hot" JS framework indeed, Vue being the "more mature but still amazing" one. But first of all, what problems are you currently facing that you need a framework to solve? What is not fine with the current codebase?

**Comment by tordans** [comment]
**Generale:**

* We need to find a way to migrate component by component; otherwise this will not get done, IMO.

* On the choice of framework: The more people know how to use the framework, the better. One goal should be to open up MC to new and future contributors.

**State management:**

A core question of mine is, how well the current custom framework stores the application state in a central place. If it does, it should be possible to migrate this state into something that can be used with React/Preact (and other frameworks) quite well. There are a few global state management libraries out there, that work well outside of react and allow to bridge frameworks (from what I understand).

Some more on this:

* Preact recently introduced Signals as State Management https://preactjs.com/blog/introducing-signals/ and https://preactjs.com/guide/v10/signals/ — Since the Signal "lives" outside the component, from my understanding it should be possible to use this as a bridge between the old and new code.
* The corresponding solutions for React would be https://jotai.org/ and https://zustand-demo.pmnd.rs/. They are very similar but different in "philosophy" ([build by the same person](https://github.com/orgs/pmndrs/repositories?q=state&type=all&language=&sort=)). I have hands on experience with Zustand, and that was great.
* In case the other options are not a good fit, we should also look at https://beta.reactjs.org/reference/react/useSyncExternalStore, which AFAIK is Reacts own hook to sync state with external libraries.
* However, even better would be to use something that keeps the state in the URL. It's not feasible to have the state in the URL _and_ a non-url-state library. I have good experience [with Tanstack Query](https://tanstack.com/router/v1/docs/guide/search-params) (former React Location). The package is in a migration phase and rename (which makes it a bit confusing), but the DX, community and authors behind it are very good. 

We should look at what we need to keep in state and check which parts belog to the URL to be shareable and which parts belog inside an "invisible" state-object.

**Tailwind CSS Eco System:**

One big goal of mine is, to switch to something that will make it a lot easier to build a great looking UI.

My tooling of choice for this is [Tailwind CSS](https://tailwindcss.com/), which we have in MapComplete as well. In addition, I would love to add components based on TailwindUI to MC. [TailwindUI](https://tailwindui.com/) are pre styled but easily customizable components (I have a paid account that I can use to contribute components). Whenever there is interaction involved, they rely on [Headless UI](https://headlessui.dev/), which is a great React Headless Library that add great interaction with great accessibility to React and is very nice to work with.

We use this combination of tools a lot at @FixMyCity and it boost our productivity a lot.

On top of that there is [Framer Motion](https://www.framer.com/motion/), which would be my goto choice once the base UI is polished and we want to improve the experience even more. There are some great videos by Sam who shows how to use Tailwind CSS an Framer Motion to create a great iOS Experience, for example https://www.youtube.com/watch?v=KzvK809rl3Q.

Having a great mobile experience would be an important follow up goal of this project, IMO.

---

Here is some input on React JS:

> Can the framework work side-by-side with the current homegrown framework? This way, the codebase can be migrated step-by-step.

Yes. It will be not the best React DX, but it is possible.

**Examples of React/Preact Integration in iD:** There was a PR https://github.com/openstreetmap/iD/pull/7861 in 2020 that added React to iD Editor. The goal was to migrate parts of the UI to React/Preact. Til specifically mentions Preact as a good and smaller alternative in https://github.com/openstreetmap/iD/issues/6540#issuecomment-662907163. (I have no hands on experience with Preact.) And Bryon mentions that – at least at that time – RapiD had React included in https://github.com/facebook/RapiD/commit/034c3316d26455a462fad876a303cf175475ccf4. This PR should also be interesting in showing how to add React to an existing framework and get the compiling and tooling (Linter etc.) going.
Maybe @til-schneider would be up to contributing his know how to this planning PR?

> Can it insert HTML-elements?
> Can it do (virtual) routing requests? (Not critical though)

We should use Tanstack Router (former name React Location) for this, see above. There are other Routers out there, but they are either tied to a React-Framework or more focussed on Page routes (the Remix router) AFAIK. Tanstack is a very good pick. It plays very nicely with Tanstack Query, which is the de facto standard to do fetching and caching of data. The goal of the new Version Tanstack Query is, to be Framework independent, so there is or will be a vanilla JS solution. We would have to check if the (currently) more stable previous release [React Location](https://react-location.tanstack.com/docs/api) also has a good solution for this.

> Can it adapt to the 'UIEventSource'-flow?

Don't know enough about that, yet. 

> How are translations handled?

There are two great packages, https://formatjs.io/docs/getting-started/installation/ (I used this most) and https://react.i18next.com/.

> Including RTL-languages?

How does this matter? Regardless, I am sure the React eco system has an answer to that. Which is a big benefit when buying into such a big eco system.

> Is loading a translation dynamically (from a theme definition) possible?

The tho packages above work differently by default, however I am sure this can be bridged with an in-between-helper.

> How good are the accessibility-features (e.g. screenreaders)? Can it be made [WCAG](https://www.w3.org/WAI/standards-guidelines/wcag/)-compatible?

This should be true for every on of our choices since all big frameworks have people working on this.
However, I really like relying on the work of other on this, since accessiblity becomes super complex, very fast.

I already mentioned Headless UI as a great tool for this. But there is also https://react-spectrum.adobe.com/react-aria/ which is a project that invests a ton of time in building great accessible components.

> How good is the tooling around the framework? Does it have a special language or is it plain typescript?

In short, its great. Prettier, Eslint, Tailwind Prettier, Testing Frameworks like [Vitest](https://vitest.dev/) and [Playwright](https://playwright.dev/) – all with great integrations into VS Code. Typescript is the de facto default for all that I liked to.

> What support does it have for MapLibre? Is a package available?

I am assuming that migrating the map itself might become one of the last steps.

However once that happens, I recommend using https://visgl.github.io/react-map-gl/. I tried adding MapLibre/Mapbox to React directly and that is certainly possible. But React Map GL handles the hard parts of rerendering and inter-library state communication and integrates with React very nicely. Our alpha Version https://radverkehrsatlas.de/regionen/bibi is also using this.

**Comment by pietervdvn** [comment]
Meeting notes 2023-01-16
With @tordans @wjtje @RobinLinde 

Frontend experience:
T: React
P, W, R: Svelte


T: Is syncing between URL-parameters and stores an issue?
P: Not really an issue, maintained by MC right now anyway

W: React is worse then Svelte
R: Svelte is better then Angular
T: I don't feel much love for react, just start with Svelte; survey indicates svelte; even if SVelte goes away, it would be easy to replace
P: don't like react as well, some experience with Svelte

What bundler to use:
Currently Parcel, very outdated 
W has a version with a new bundler and svelte; has been added to the contributors of the repo


As most are experienced with svelte, maybe pick Svelte and test a component to port first?
T: migrate something simple, does it still work? Are there rendering issues? To many rerenders?
   Take a part that is working fine now
P: proposal: home screen or panel selecting layers, filters and background 
T: Background/layer panel probably best component

W: push build with vite (dev works, build still fails)
R: build with netlify/vercel (netlify has opensource plan with conditions)
P: maybe time to get a (build)-server

# Organisation

Visual design part: T talks to Heiko, tries to get a feel if and when to contribute e.g. basestyles for buttons.

T: Meeting every two weeks to sync and see where to contribute; (not on Tuesday)
P: Probably working Wed, Thu, Fr on MapComplete, so maybe Wednesdays? 14:00

**Next meeting: wednesday 01/02, 14:00**

**Comment by pietervdvn** [comment]
As the framework is decided on, this issue can be closed.
